File: Lua/Framework/Manager/Wrapper/PowerManager.lua
local Delegate = require("Delegate")
local EventConst = require("EventConst")
local ModuleRefer = require("ModuleRefer")
local I18N = require("I18N")

local BaseManager = require("BaseManager")

---@class PowerManager:BaseManager
---@field new fun():PowerManager
---@field super BaseManager
local PowerManager = class('PowerManager', BaseManager)

function PowerManager:ctor()
    ---@type CS.DragonReborn.Utilities.PowerManager
    self.csManager = CS.DragonReborn.Utilities.PowerManager.Instance
    self.csManager:OnGameInitialize()
end

function PowerManager:GetCurrentThermalStatus()
    return self.csManager:GetCurrentThermalStatus()
end

function PowerManager:IsPowerSaveMode()
    return self.csManager:IsPowerSaveMode()
end

function PowerManager:KeepScreenOn()
    self.csManager:KeepScreenOn()
end

function PowerManager:NoKeepScreenOn()
    self.csManager:NoKeepScreenOn()
end

function PowerManager:BackupAndroidWindowFlags()
    self.csManager:BackupAndroidWindowFlags()
end

function PowerManager:RestoreAndroidWindowFlags()
    self.csManager:RestoreAndroidWindowFlags()
end

function PowerManager:Reset()
    self.csManager:Reset()
end

function PowerManager:OnLowMemory()
    try_catch_traceback_with_vararg(self.csManager.OnLowMemory, nil, self.csManager)
end

return PowerManager

File: Assets/__Script/Utilities/PowerManager.cs

// ReSharper disable once CheckNamespace
namespace DragonReborn.Utilities
{
	public class PowerManager : Singleton<PowerManager>, IManager
	{
		private bool _init;
		private IFrameworkNativePowerManager _nativePowerManager;

		// for old luac compatible
		public void Initialize()
		{
			// InitOnce();
		}

		public void OnGameInitialize(object configParam)
		{
			InitOnce();
		}

		public void Reset()
		{
			_init = false;
			_nativePowerManager?.Reset();
			_nativePowerManager = null;
		}

		private void InitOnce()
		{
			if (_init) return;
			_init = true;
			FrameworkInterfaceManager.QueryFrameInterface(out _nativePowerManager);
			_nativePowerManager?.Init();
		}

		public NativePowerThermalStatus GetCurrentThermalStatus()
		{
			return _nativePowerManager?.GetCurrentThermalStatus() ?? NativePowerThermalStatus.THERMAL_STATUS_UNKNOWN;
		}

		public bool IsPowerSaveMode()
		{
			return _nativePowerManager?.IsPowerSaveMode() ?? false;
		}

		public void KeepScreenOn()
		{
			_nativePowerManager?.KeepScreenOn();
		}
		
		public void NoKeepScreenOn()
		{
			_nativePowerManager?.NoKeepScreenOn();
		}
		
		public void BackupAndroidWindowFlags()
		{
			_nativePowerManager?.BackupAndroidWindowFlags();
		}

		public void RestoreAndroidWindowFlags()
		{
			_nativePowerManager?.RestoreAndroidWindowFlags();
		}

		public void OnLowMemory()
		{

		}
	}
}
